components/prism-cmake.js

```diff
@@ -0,0 +1,8 @@
+    "function": /\b(add_compile_definitions|add_compile_options|add_custom_command|add_custom_target|add_definitions|add_dependencies|add_executable|add_library|add_link_options|add_subdirectory|add_test|aux_source_directory|break|build_command|build_name|cmake_host_system_information|cmake_minimum_required|cmake_parse_arguments|cmake_policy|configure_file|continue|create_test_sourcelist|ctest_build|ctest_configure|ctest_coverage|ctest_empty_binary_directory|ctest_memcheck|ctest_read_custom_files|ctest_run_script|ctest_sleep|ctest_start|ctest_submit|ctest_test|ctest_update|ctest_upload|define_property|else|elseif|enable_language|enable_testing|endforeach|endfunction|endif|endmacro|endwhile|exec_program|execute_process|export|export_library_dependencies|file|find_file|find_library|find_package|find_path|find_program|fltk_wrap_ui|foreach|function|get_cmake_property|get_directory_property|get_filename_component|get_property|get_source_file_property|get_target_property|get_test_property|if|include|include_directories|include_external_msproject|include_guard|include_regular_expression|install|install_files|install_programs|install_targets|link_directories|link_libraries|list|load_cache|load_command|macro|make_directory|mark_as_advanced|math|message|option|output_required_files|project|qt_wrap_cpp|qt_wrap_ui|remove|remove_definitions|return|separate_arguments|set|set_directory_properties|set_property|set_source_files_properties|set_target_properties|set_tests_properties|site_name|source_group|string|subdir_depends|subdirs|target_compile_definitions|target_compile_features|target_compile_options|target_include_directories|target_link_directories|target_link_libraries|target_link_options|target_sources|try_compile|try_run|unset|use_mangled_mesa|utility_source|variable_requires|variable_watch|while|write_file)\b/,
```

Comment:
Guessing from the test, could it be that a function is always followed by `(`?

If so, we can simplify this expression to `/\b[a-z_]\w*(?=\s*\()/i`.

Reply:
Unfortunately, it's not that simple. if() is also followed by (. This is the list of built-in functions generated by CMake's help. You can define your own functions, but it's quite rare and there is no way to know it with a regex.

Reply:
But `if` is already matched as a `keyword`, so it can't be matched again as a `function`.
Other programming languages (every C-style one) have the same issue with `if`, `while`, etc. and they use this exact trick.

So after matching the keywords, are there any function-like constructs left?

Reply:
Wait. `if` is also in the functions list.

Reply:
Welcome to CMake world ;)

Reply:
In general constructs like ifs and whiles are used very sparingly during CMake "development" so treating them like keywords or functions does not really matter. I think it's enough they are "recognized" by the highlighter and coloured in some way.

Reply:
I just think that these, well, _feel_ like keywords, so I would much rather have them highlighted as such. `function` could then match user-defined functions and macros using the pattern: `/\b[a-z_]\w*(?=\s*\()/i`.

What do you think?

Reply:
I agree that user-defined functions should be highlighted differently. However e.g. WIN32 should be highlighted differently than add_library. All of the IDEs supporting CMake work this way and it makes reading code easier.

I propose to use the `keyword` highlighting for `add_library, find_package, if, while` - everything that's generally known to CMake and written with lowercase with parens after.

I agree with the function proposal of yours. Let's use it.

Let's leave properties as they are.

Now, what can we do with the current `keyword` marking? It's different than the others. Can we use some other class?

Reply:
Looks good!
As suggested [here](https://github.com/PrismJS/prism/pull/1820#issuecomment-474110085) we could  the current `keyword`

Reply:
As you pointed out: "We could give the variables list the `variable` highlighting as well."
The variable concept matches what's underneath the current `keyword` as e.g. `CMAKE_ABSOLUTE_DESTINATION_FILES` is a variable holding a list of files and can be used using variable semantics.

Reply:
I've pushed my proposal.
It lacks function tag as I tried the proposed one and it did not match what I expected i.e. `bar` in expression `bar()`.

Reply:
Strange, according to Regexr, your regex works just fine. I will debug this a bit more.

Reply:
That because you already match the `(` in `punctuation`. Is usually a good idea to match operators and punctuation last.

Reply:
Yes, you are right.
I've pushed the correction.